docs: Add Zod usage guideline (#25080)
renovatebot/renovate: docs/development/zod.md

This document describes some guides as to how and why to use Zod features.

The key concept of schema validation is to find the right balance between strictness of contracts between separately developed systems and the permissiveness of the Renovate itself.
We want Renovate to be only as strict as it needs to be (e.g. about optional fields which a public registry might always have but self-hosted implementations may leave off) but not to make _assumptions_ about the presence of fields which could lead to run-time errors when they are missing.

## When/where to use Zod

We should use Zod for validating:

- Data received from external APIs and data sources, particularly the `lib/modules/datasource/*` section of Renovate
- Data parsed from files in the repository, particularly the `lib/modules/manager/*` section of Renovate

The `cdnjs` datasource is a good example of using Zod schema validations on API responses from external sources.

The `composer` manager is a good example of using Zod schema validation in a manager to validate parsed files in a repository.

## Technical guide

### Use schema.ts files for Zod schemas

> Try to locate/isolate Zod schemas in their own schema.ts files to keep them organized and reusable. [TODO: do our examples follow that?]

Reviewer: We need to pick other examples or _improve_ our examples if not

Author: I've added a paragraph stating something like "while it's the general rule, sometimes it can be broken when it makes sense"

Reviewer: Can't we get _one_ example?

This could lead to cases when Renovate could've continued processing, but didn't.

Remember: our goal is not to validate that data corresponds to any official specifications, but rather to ensure that the data is enough for Renovate to use.

There are some techniques to make it more permissive to the input data.

#### Use `.catch()` to force default values

```ts
const Box = z.object({
  width: z.number().catch(10),
  height: z.number().catch(10),
});

const box = Box.parse({ width: 10, height: null });
// => { width: 10, height: 10 }
```

#### Use `LooseArray` and `LooseRecord` to filter out incorrect values from collections

Suppose you want to validate an array and retain only number values in it.
By using only methods provided by `zod` library, you'll have to write something like this:

```ts
const OnlyNumbers = z
  .array(z.union([z.number(), z.null()]).catch(null))
  .transform((xs) => xs.filter((x): x is number => x !== null));
```

While the problem is common, the code is quite complicated.

Instead, you should use `LooseArray` and `LooseRecord` helpers provided in `schema-utils.ts`:

```ts
const OnlyNumbers = LooseArray(z.number());
```

> [TODO: more details on how the above would be used]

Reviewer: OnlyNumbers would need to be used somewhere else to be useful, right?

Author: This is just an artificial example, I used it instead of something real but more verbose, like `ReleaseResult` filtering or something

Reviewer: I found the example to be unhelpful because it was too artificial. Can you create a more realistic one?

[The `composer` manager](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/composer/schema.ts) is a good example of using Zod schema validation in a manager to validate parsed files in a repository.

## Technical guide

### Use `schema.ts` files for Zod schemas

Following well-known refactoring principles, we should locate schemas accordingly.
Generally, we keep them in the `schema.ts` files, and even better if we test them in the `schema.spec.ts` files.
However, for small usage areas, it's fine to keep them in the same file for reader's convenience.

Sometimes, we create or extend schemas on the fly, but it should be avoided as it has performance implications.

### Name schemas without any `Schema` suffix

Schemas are distinguished by being named with capital letter:

```ts
const ComplexNumber = z.object({
  re: z.number(),
  im: z.number(),
});
```

Don't use names like `ComplexNumberSchema` for that.

### Inferred types

Create inferred types after schemas if they're needed somewhere in the code.
Place such inferred types just after the schema definition using the same name.

> While text editors may confuse schema and type name sometimes, it's obvious which is which from the syntactic context.

Author: Okay, just realized all these are "text editors" in my head for quite a while

Example:

```ts
export const User = z.object({
  firstName: z.string(),
  lastName: z.string(),
});
export type User = z.infer<typeof User>;
```

### Specify only necessary fields

> External data being queried by Renovate can be very complex, while we may only need a subset of fields for our use cases.

Reviewer 2: ```suggestion
The external data that Renovate queries can be very complex, but Renovate may only need some of those fields.
```

> Avoid over-specifying schemas, and instead extract only the minimum necessary fields.

Reviewer 2: ```suggestion
Avoid over-specifying schemas, only extract fields Renovate really needs.
```

> If we don't include them, the surface of the contract between external data source is minimal, meaning less errors to fix in the future.

Reviewer 2: ```suggestion
This reduces the "surface of the contract" between the external data source and Renovate, which means less errors to fix in the future.
```

> Example of **incorrect** usage if we only care about width, height and length of a box:

Reviewer 2: ```suggestion
For example, say you want Renovate to know about the width, height and length of a box.
You should _avoid_ code like this:
```

```ts
const Box = z.object({
  width: z.number(),
  height: z.number(),
  length: z.number(),
  color: z.object({
    red: z.number(),
    green: z.number(),
    blue: z.number(),
  })
  weight: z.number(),
});

const { width, height, length } = Box.parse(input);
const volume = width * height * length;
```

> Example of **correct** usage:

Reviewer 2: ```suggestion
The code above refers to the `color` and `weight` fields, which Renovate does _not_ need to do its job.
Here's the **correct** code:
```

```ts
const Box = z.object({
  width: z.number(),
  height: z.number(),
  length: z.number(),
});

const { width, height, length } = Box.parse(input);
const volume = width * height * length;
```

### Use `Json`, `Yaml` and `Toml` for string parsing

> Sometimes we need to perform additional step such as `JSON.parse()` before validation of the data structure.

Reviewer 2: ```suggestion
You may need to perform extra steps like `JSON.parse()` before you can validate the data structure.
```

> Use helpers from `schema-utils.ts` for this purpose.

Reviewer 2: ```suggestion
Use the helpers in `schema-utils.ts` for this purpose.
```

> Here is an **incorrect** way to parse from string:

Reviewer 2: ```suggestion
The **wrong** to parse from string:
```

```ts
try {
  const json = JSON.parse(input);
  return ApiResult.parse(json);
} catch (e) {
  return null;
}
```

The **correct** way to parse from string:

```ts
Json.pipe(ApiResult).parse(input);
```

### Use `.transform()` method to process parsed data

Schema validation helps to be more confident with types during downstream data transformation.

You can go even further and perform some transformations as the part of schema itself.

This is an example of **undesired** data transformation:

```ts
const Box = z.object({
  width: z.number(),
  height: z.number(),
  length: z.number(),
});

const { width, height, length } = Box.parse(input);
const volume = width * height * length;
```

> Instead, use more idiomatic `.tranform()` method:

Reviewer 2: ```suggestion
Instead, use the idiomatic `.tranform()` method:
```

```ts
const Volume = z
  .object({
    width: z.number(),
    height: z.number(),
    length: z.number(),
  })
  .transform(({ width, height, length }) => width * height * length);

Volume.parse({
  width: 10,
  height: 20,
  length: 125,
});
```

### Separate validation from transformation

> When parsing third party data, we are typically doing the following:

Reviewer 2: ```suggestion
When parsing third-party data, we are typically doing the following:
```

> - Validating that the data is correct/sufficient to use
> - Transforming it into a standardized format for internal usage

Reviewer 2: ```suggestion
1. Validating the third-party data is correct/sufficient
1. Transforming the third-party data into a format for internal use
```

> Although it's not a strict requirement, your code will be cleaner if you perform the validation step first and then follow with transformation.

Reviewer 2: ```suggestion
Write clean code, and perform the validation step first, and follow up with the transformation.
```

It's a bit weird to say: "Here's how you write clean code. But you don't really need to write clean code right after." So how about this instead?

### Stick to permissive behavior when possible

> Zod schemas are strict, and even some insufficient field is incorrect, the whole data will be treated like malformed.
> This could lead to cases when Renovate could've continued processing, but didn't.

Reviewer 2: ```suggestion
Zod schemas are strict, which means that if some field is wrong, or missing data, then the whole dataset is considered malformed.
Because Renovate uses Zod, it would then abort processing, even if we want Renovate to continue processing!
```

> Remember: our goal is not to validate that data corresponds to any official specifications, but rather to ensure that the data is enough for Renovate to use.

Reviewer 2: ```suggestion
Remember: we want to make sure the incoming data is good enough for Renovate to work.
We do _not_ need to validate the data corresponds to any official specification.
```

> There are some techniques to make it more permissive to the input data.

Reviewer 2: ```suggestion
Here are some techniques to make Zod more permissive about the input data.
```

#### Use `.catch()` to force default values

```ts
const Box = z.object({
  width: z.number().catch(10),
  height: z.number().catch(10),
});

const box = Box.parse({ width: 10, height: null });
// => { width: 10, height: 10 }
```

#### Use `LooseArray` and `LooseRecord` to filter out incorrect values from collections

> Suppose you want to validate an array and retain only number values in it.

Reviewer 2: ```suggestion
Say you want to validate an array and keep the number values from the array.
```

> By using only methods provided by `zod` library, you'll have to write something like this:

Reviewer 2: ```suggestion
If you only use methods from the `zod` library, you would need to write something like this:
```

```ts
const OnlyNumbers = z
  .array(z.union([z.number(), z.null()]).catch(null))
  .transform((xs) => xs.filter((x): x is number => x !== null));
```

> While the problem is common, the code is quite complicated.

Reviewer 2: ```suggestion
You're working on a common problem, but end up writing complicated code.
```

> Instead, you should use `LooseArray` and `LooseRecord` helpers provided in `schema-utils.ts`:

Reviewer 2: ```suggestion
Instead, you should use the `LooseArray` and `LooseRecord` helpers from `schema-utils.ts` to write simpler code:
```

```ts
const OnlyNumbers = LooseArray(z.number());

const xs = OnlyNumbers.parse([1, 2, null, 3, 'foobar', 4]);
// => [1, 2, 3, 4]
```

### Combining with `Result` class

> Class `Result` (and also `AsyncResult`) represents the result of an operation, e.g. `Result.ok(200)` or `Result.err(404)`.

Reviewer 2: ```suggestion
The `Result` (and `AsyncResult`) class represents the result of an operation, like `Result.ok(200)` or `Result.err(404)`.
```

> It supports `.transform()` method, which is similar to `zod`'s one.

Reviewer 2: ```suggestion
It supports the `.transform()` method, which is similar to `zod`'s.
```

> It also supports `.onResult()` and `.onError()` methods for side-effectful result inspection.

Reviewer 2: "Side-effectful result inspection" means nothing to me. Are you sure the developers reading this know what this means? I would prefer some simpler words here. 😄

Reviewer: I think we're ok to leave it in

> Once all result manipulations are done, you may call `.unwrap()`, `.unwrapOrElse()` or `.unwrapOrThrow()` methods to obtain the underlying result value.

Reviewer 2: ```suggestion
After all result manipulations are done, you may call `.unwrap()`, `.unwrapOrElse()` or `.unwrapOrThrow()` methods to get the underlying result value.
```

> You can wrap schema parsing result into the `Result` class:

Reviewer 2: ```suggestion
You can wrap the schema parsing result into the `Result` class:
```

```ts
const { val, err } = Result.parse(url, z.string().url())
  .transform((url) => http.get(url))
  .onError((err) => {
    logger.warn({ err }, 'Failed to fetch something important');
  })
  .transform((res) => res.body);
```

> You can use schema parsing in the middle of `Result` transform chain:

Reviewer 2: ```suggestion
You can use schema parsing in the middle of the `Result` transform chain:
```

```ts
const UserConfig = z.object({
  /* ... */
});

const config = await Result.wrap(readLocalFile('config.json'))
  .transform((content) => Json.pipe(UserConfig).safeParse(content))
  .unwrapOrThrow();
```

### Combining with `Http` class

> `Http` class supports schema validation for the results of JSON family of methods, e.g. `.getJson()`, `.postJson()`, etc.

Reviewer 2: ```suggestion
`Http` class supports schema validation for the results of the "JSON family of methods" like `.getJson()`, `.postJson()`, etc.
```

Under the hood, `.parseAsync()` method is used (**important consequence**: in case of invalid data, it will throw).

> Place schema as the last argument of the method:

Reviewer 2: ```suggestion
The schema should be the _last_ argument of the method:
```

Author: ```suggestion
Provide schema in the last argument of the method:
```

```ts
const Users = z.object({
  users: z.object({
    id: z.number(),
    firstName: z.string(),
    lastName: z.string(),
  }),
});

const { body: users } = await http.getJson(
  'https://dummyjson.com/users',
  LooseArray(User)
);
```

> For GET requests, use `.getJsonSafe()` method which returns `Result` instance:

Reviewer 2: ```suggestion
For GET requests, use the `.getJsonSafe()` method which returns a `Result` instance:
```

> ### Table of Contents
> 
> - [Zod schema guideline](#zod-schema-guideline)
>   - [When and where to use Zod](#when-and-where-to-use-zod)
>   - [Technical guide](#technical-guide)
>     - [Use `schema.ts` files for Zod schemas](#use-schemats-files-for-zod-schemas)
>     - [Name schemas without any `Schema` suffix](#name-schemas-without-any-schema-suffix)
>     - [Inferred types](#inferred-types)
>     - [Specify only necessary fields](#specify-only-necessary-fields)
>     - [Use `Json`, `Yaml` and `Toml` for string parsing](#use-json-yaml-and-toml-for-string-parsing)
>     - [Use `.transform()` method to process parsed data](#use-transform-method-to-process-validated-data)
>       - [Rename and move fields at the top level transform](#rename-and-move-fields-at-the-top-level-transform)
>     - [Stick to permissive behavior when possible](#stick-to-permissive-behavior-when-possible)
>       - [Use `.catch()` to force default values](#use-catch-to-force-default-values)
>       - [Use `LooseArray` and `LooseRecord` to filter out incorrect values from collections](#use-loosearray-and-looserecord-to-filter-out-incorrect-values-from-collections)
>     - [Combining with `Result` class](#combining-with-result-class)
>     - [Combining with `Http` class](#combining-with-http-class)

Reviewer 2: ```suggestion
If you're viewing this file on GitHub, you can use a special button to see a auto-generated table of contents.
Read the [GitHub Docs, Basic writing and formatting syntax, headings](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#headings) section to learn more about this button.
```

The GitHub Docs shows what the "table of contents button and icon" looks like. So we can drop the manual TOC and rely on GitHub's TOC.

I don't know if we need to say: "If you're using a text editor you can probably get a table of contents or overview tab." It feels too obvious to mention this.

Author: I'm afraid that almost nobody actually uses it. I noticed this feature right in the moment of me searching how to make Markdown TOC for GitHub.

### Table of Contents

- [Zod schema guideline](#zod-schema-guideline)
  - [When and where to use Zod](#when-and-where-to-use-zod)
  - [Technical guide](#technical-guide)
    - [Use `schema.ts` files for Zod schemas](#use-schemats-files-for-zod-schemas)
    - [Name schemas without any `Schema` suffix](#name-schemas-without-any-schema-suffix)
    - [Inferred types](#inferred-types)
    - [Specify only necessary fields](#specify-only-necessary-fields)
    - [Use `Json`, `Yaml` and `Toml` for string parsing](#use-json-yaml-and-toml-for-string-parsing)
    - [Use `.transform()` method to process parsed data](#use-transform-method-to-process-validated-data)
      - [Rename and move fields at the top level transform](#rename-and-move-fields-at-the-top-level-transform)
    - [Stick to permissive behavior when possible](#stick-to-permissive-behavior-when-possible)
      - [Use `.catch()` to force default values](#use-catch-to-force-default-values)
      - [Use `LooseArray` and `LooseRecord` to filter out incorrect values from collections](#use-loosearray-and-looserecord-to-filter-out-incorrect-values-from-collections)
    - [Combining with `Result` class](#combining-with-result-class)

>     - [Combining with `Http` class](#combining-with-http-class)

Reviewer: Is this ToC really needed? I think it risks getting out of date, makes it harder to refactor/update, etc. I would only like it if it can be automated

Author: Let's remove it when we notice it's outdated the second or third time? Links are checked by CI, and mantaining it should not that hard compared to spending time to automate it.

IMO yes, TOC is useful, because not too much people may know GitHub has automatic TOC dropdown in the Markdown files.
